File: Tutor Demos/Week 5/Java/Customer.java
public class Customer {
    public static void main(String[] args) {
        new Customer().use();
    }
    private final Account savingsAccount;
    private final Account loanAccount;

    public Customer() {
        savingsAccount = new Account(AccountType.SAVINGS);
        loanAccount = new Account(AccountType.LOAN);
    }

    public void use() {
        char choice;
        while ((choice = readChoice()) != 'x') {
            switch (choice) {
                case 'd': deposit(); break;
                case 'w': withdraw(); break;
                case 't': transfer(); break;
                case 's': show(); break;
                default: help(); break;
            }
        }
        System.out.println("Done");
    }

    private char readChoice() {
        System.out.print("Customer menu (d/w/t/s/x): ");
        return In.nextChar();
    }

    private void deposit() {
        double amount = readAmount("deposit");
        savingsAccount.deposit(amount);
    }

    private void withdraw() {
        double amount = readAmount("withdraw");
        if (savingsAccount.has(amount))
            savingsAccount.withdraw(amount);
        else
            System.out.println("Insufficient funds");
    }

    private void transfer() {
        double amount = readAmount("transfer");
        if (savingsAccount.has(amount))
            savingsAccount.transferTo(amount, loanAccount);
        else
            System.out.println("Insufficient funds");
    }

    private double readAmount(String action) {
        System.out.print("Amount to " + action + ": $");
        return In.nextDouble();
    }

    private void show() {
        System.out.println(savingsAccount);
        System.out.println(loanAccount);
    }

    private void help() {
        System.out.println("Menu options");
        System.out.println("d = deposit");
        System.out.println("w = withdraw");
        System.out.println("t = transfer");
        System.out.println("s = show");
        System.out.println("x = exit");
    }
}

File: Tutor Demos/Week 5/customer.py
from account import Account

class Customer:
    def __init__(self):
        self.savings_account = Account("Savings")
        self.loan_account = Account("Loan")

    def use(self):
        choice = ''
        while (choice := input("Customer menu (d/w/t/s/x): ")) != 'x':
            if choice == 'd':
                self.deposit()
            elif choice == 'w':
                self.withdraw()
            elif choice == 't':
                self.transfer()
            elif choice == 's':
                self.show()
            else:
                self.help()
        print("Done")

    #Not strictly necessary in python, but a very reused and error prone piece of code
    #Good to have only 1 instance of it
    def read_amount(self, action):
        return float(input(f"Amount to {action}: $"))

    def deposit(self):
        amount = self.read_amount("deposit")
        self.savings_account.deposit(amount)

    def withdraw(self):
        amount = self.read_amount("withdraw")
        if self.savings_account.has(amount):
            self.savings_account.withdraw(amount)
        else:
            print("Insufficient funds")

    def transfer(self):
        amount = self.read_amount("withdraw")
        if self.savings_account.has(amount):
            self.savings_account.transfer_to(amount, self.loan_account)
        else:
            print("Insufficient funds")

    def show(self):
        print(self.savings_account.__str__())
        print(self.loan_account.__str__())

    def help(self):
        print("Menu options")
        print("d = deposit")
        print("w = withdraw")
        print("t = transfer")
        print("s = show")
        print("x = exit")

if __name__ == "__main__":
    Customer().use()